File: extension/src/tools/quizzes.ts
/**
 * list_quizzes — metadata only, deliberately.
 * Ported from src/avenue_mcp/tools/quizzes.py.
 *
 * Names, dates, availability windows, and whether you've attempted. Quiz
 * QUESTIONS and ANSWERS are out of scope and stay that way: knowing a quiz is
 * due Friday is calendar information; retrieving its contents while it is open
 * is the thing academic-integrity policies exist to prohibit.
 *
 * On this instance `quizzes/{id}/attempts/` is 403 (docs/08), so attempt counts
 * come back null rather than zero — "I don't know" and "you haven't taken it"
 * are different answers.
 */

import { avenue } from "../avenue/client.js";
import { AvenueError } from "../avenue/errors.js";
import { daysUntil, describe, nowUtc, parseD2L } from "../avenue/dates.js";
import { asFloat, asInt, isObj, pick } from "../avenue/models.js";
import { calendarEvents } from "./calendar.js";
import { courseName } from "./context.js";

export async function listQuizzes(args: {
  org_unit_id: number;
  include_completed?: boolean;
}) {
  const { org_unit_id } = args;
  const includeCompleted = args.include_completed ?? true;
  const now = nowUtc();

  let raw: unknown[];
  try {
    raw = await avenue.getPaged("le", `${org_unit_id}/quizzes/`);
  } catch (err) {
    if (err instanceof AvenueError && err.kind === "PermissionDenied") {
      return calendarFallback(org_unit_id, err.message);
    }
    throw err;
  }

  const quizzes = [];
  for (const q of raw) {
    if (!isObj(q)) continue;
    const quizId = asInt(pick(q, "QuizId", "Id"));
    if (quizId === null) continue;

    const start = parseD2L(pick(q, "StartDate", "AvailabilityStartDate"));
    const due = parseD2L(pick(q, "DueDate", "Due"));
    const end = parseD2L(pick(q, "EndDate", "AvailabilityEndDate"));

    const { used, best } = await attempts(org_unit_id, quizId);

    // Three dates, three meanings. A quiz can be submittable AFTER due but
    // BEFORE end — telling a student it is closed when it is merely late is as
    // damaging as the reverse.
    const openNow = (!start || start <= now) && (!end || now <= end);
    const pastDueOpen = Boolean(due && now > due && (!end || now <= end));

    const status = used === null ? "unknown" : used > 0 ? "attempted" : "not_attempted";
    if (status === "attempted" && !includeCompleted) continue;

    quizzes.push({
      quiz_id: quizId,
      name: String(pick(q, "Name", "Title") ?? ""),
      start_date: describe(start),
      due_date: describe(due),
      end_date: describe(end),
      days_until_due: daysUntil(due, now),
      attempts_allowed: asInt(pick(q, "AttemptsAllowed", "MaxAttempts", "NumberOfAttemptsAllowed")),
      attempts_used: used,
      status,
      is_available_now: openNow,
      is_past_due_but_open: pastDueOpen,
      is_closed: Boolean(end && now > end),
      best_score: best,
    });
  }

  quizzes.sort(
    (a, b) =>
      (a.due_date?.utc ?? "9999").localeCompare(b.due_date?.utc ?? "9999") ||
      a.name.localeCompare(b.name),
  );

  const statusUnknown = quizzes.some((q) => q.status === "unknown");

  return {
    org_unit_id,
    course_name: await courseName(org_unit_id),
    quizzes,
    count: quizzes.length,
    degraded: false,
    attempt_status_available: !statusUnknown,
    note:
      "Quiz metadata only — questions and answers are never retrieved." +
      (statusUnknown
        ? " Attempt status is unavailable on this instance, so status is 'unknown'. Do not tell the user whether they have taken a quiz; point them at Avenue."
        : ""),
  };
}

/**
 * Attempt count and best score.
 *
 * Returns null (not 0) when the route is denied. Reporting zero attempts on
 * silence would tell a student they haven't taken a quiz they may well have.
 */
async function attempts(
  orgUnitId: number,
  quizId: number,
): Promise<{ used: number | null; best: number | null }> {
  let data: unknown[];
  try {
    data = await avenue.getPaged("le", `${orgUnitId}/quizzes/${quizId}/attempts/`);
  } catch {
    return { used: null, best: null };
  }

  let used = 0;
  let best: number | null = null;
  for (const att of data) {
    if (!isObj(att)) continue;
    used++;
    const score = asFloat(pick(att, "Score", "PointsEarned"));
    if (score !== null && (best === null || score > best)) best = score;
  }
  return { used, best };
}

/** Deadlines survive a blocked quizzes route; attempt status does not. */
async function calendarFallback(orgUnitId: number, reason: string) {
  let events: Awaited<ReturnType<typeof calendarEvents>> = [];
  try {
    events = await calendarEvents(orgUnitId);
  } catch {
    events = [];
  }

  const quizzes = events
    .filter((ev) => ev.kind === "quiz")
    .map((ev) => ({
      quiz_id: null,
      name: ev.title,
      due_date: ev.due_date,
      days_until_due: daysUntil(ev.dueAt),
      attempts_allowed: null,
      attempts_used: null,
      status: "unknown" as const,
      is_available_now: null,
      is_past_due_but_open: null,
    }));

  return {
    org_unit_id: orgUnitId,
    course_name: await courseName(orgUnitId),
    quizzes,
    count: quizzes.length,
    degraded: true,
    attempt_status_available: false,
    note:
      "The quizzes API is not available to student accounts on this instance, so " +
      "attempt status and availability windows are unknown. These entries come " +
      `from the course calendar. (${reason})`,
  };
}

File: src/avenue_mcp/tools/quizzes.py
"""list_quizzes -- metadata only, deliberately.

Names, dates, availability windows, and whether you've attempted. Quiz
*questions* and *answers* are out of scope and stay that way: knowing a quiz is
due Friday is calendar information; retrieving its contents while it is open is
the thing academic integrity policies exist to prohibit.
"""

from __future__ import annotations

import logging
from typing import Any

from avenue_mcp.client import models as m
from avenue_mcp.context import AppContext
from avenue_mcp.errors import APIError, PermissionDeniedError
from avenue_mcp.util.capability import capability_status, describe_denial
from avenue_mcp.util.dates import days_until, describe, now_utc, parse_d2l

log = logging.getLogger(__name__)


async def list_quizzes(
    ctx: AppContext, org_unit_id: int, include_completed: bool = True
) -> dict[str, Any]:
    await ctx.require_session()
    tz = ctx.settings.timezone
    now = now_utc()

    try:
        raw = await ctx.client.get_paged("le", f"{org_unit_id}/quizzes/")
    except (PermissionDeniedError, APIError) as exc:
        log.info("quizzes route unavailable for %s: %s", org_unit_id, exc)
        return await _calendar_fallback(ctx, org_unit_id, str(exc))

    quizzes: list[dict[str, Any]] = []
    attempts_unavailable = False
    for q in raw:
        if not isinstance(q, dict):
            continue
        quiz_id = m.as_int(m.pick(q, "QuizId", "Id"))
        if quiz_id is None:
            continue

        start = parse_d2l(m.pick(q, "StartDate", "AvailabilityStartDate"))
        due = parse_d2l(m.pick(q, "DueDate", "Due"))
        end = parse_d2l(m.pick(q, "EndDate", "AvailabilityEndDate"))

        attempts_allowed = m.as_int(
            m.pick(q, "AttemptsAllowed", "MaxAttempts", "NumberOfAttemptsAllowed")
        )
        attempts_used, best = await _attempts(ctx, org_unit_id, quiz_id)
        if attempts_used is None:
            # Denied, not zero. Saying "not_attempted" about a quiz the student
            # actually wrote is the same confidently-wrong claim the assignment
            # path guards against.
            status = "unknown"
            attempts_unavailable = True
        else:
            status = "attempted" if attempts_used > 0 else "not_attempted"

        # Three dates, three meanings. A quiz can be submittable AFTER due but
        # BEFORE end -- telling a student it is closed when it is merely late is
        # as damaging as the reverse.
        open_now = (start is None or start <= now) and (end is None or now <= end)
        past_due_open = bool(due and now > due and (end is None or now <= end))

        record = {
            "quiz_id": quiz_id,
            "name": m.pick(q, "Name", "Title"),
            "start_date": describe(start, tz),
            "due_date": describe(due, tz),
            "end_date": describe(end, tz),
            "days_until_due": days_until(due, now),
            "attempts_allowed": attempts_allowed,
            "attempts_used": attempts_used,
            "status": status,
            "is_available_now": open_now,
            "is_past_due_but_open": past_due_open,
            "is_closed": bool(end and now > end),
            "best_score": best,
        }
        if record["status"] == "attempted" and not include_completed:
            continue
        quizzes.append(record)

    quizzes.sort(key=lambda x: (x["due_date"]["utc"] or "9999", x["name"] or ""))

    note = "Quiz metadata only -- questions and answers are never retrieved."
    out: dict[str, Any] = {
        "org_unit_id": org_unit_id,
        "course_name": await ctx.course_name(org_unit_id),
        "quizzes": quizzes,
        "count": len(quizzes),
        "degraded": False,
        "attempt_status_available": not attempts_unavailable,
    }
    if attempts_unavailable:
        out["capability_status"] = capability_status(ctx.settings, "quiz_attempts")
        note += (
            " Attempt status is NOT available -- the quiz attempts route was "
            "denied, so every quiz shows status 'unknown'. "
            + describe_denial(ctx.settings, "quiz_attempts")
            + " Do not tell the user whether they have taken a quiz; point them "
            "at Brightspace. Dates and availability above are accurate."
        )
    out["note"] = note
    return out


async def _attempts(
    ctx: AppContext, org_unit_id: int, quiz_id: int
) -> tuple[int | None, float | None]:
    """`None` for the count means UNREADABLE, not zero.

    The attempts route is denied to students on both measured instances, so
    conflating "denied" with "no attempts" makes every quiz report
    `not_attempted` -- including a midterm the student demonstrably wrote.
    """
    try:
        data = await ctx.client.get_paged(
            "le", f"{org_unit_id}/quizzes/{quiz_id}/attempts/"
        )
    except APIError:
        return None, None

    count = 0
    best: float | None = None
    for att in data:
        if not isinstance(att, dict):
            continue
        count += 1
        score = m.as_float(m.pick(att, "Score", "PointsEarned"))
        if score is not None and (best is None or score > best):
            best = score
    return count, best


async def _calendar_fallback(
    ctx: AppContext, org_unit_id: int, reason: str
) -> dict[str, Any]:
    """Deadlines survive a blocked quizzes route; attempt status does not."""
    from avenue_mcp.tools.assignments import _calendar_events

    try:
        events = await _calendar_events(ctx, org_unit_id)
    except APIError:
        events = []

    quizzes = [
        {
            "quiz_id": None,
            "name": ev["title"],
            "due_date": ev["due_date"],
            "days_until_due": days_until(ev["due_dt"]),
            "attempts_allowed": None,
            "attempts_used": None,
            "status": "unknown",
            "is_available_now": None,
            "is_past_due_but_open": None,
        }
        for ev in events
        if ev["kind"] == "quiz"
    ]

    return {
        "org_unit_id": org_unit_id,
        "course_name": await ctx.course_name(org_unit_id),
        "quizzes": quizzes,
        "count": len(quizzes),
        "degraded": True,
        "note": (
            "The quizzes API is not available to student accounts on this "
            "instance, so attempt status and availability windows are unknown. "
            f"These entries come from the course calendar. ({reason})"
        ),
    }
